File: group 3/variant11/SubscriptionService.cs
using System;

namespace UniversityTesting.Variant11
{
    public sealed class SubscriptionService
    {
        public sealed record Request(int Users, int Months, bool NonprofitClaimed, bool AutoRenew);

        public sealed record Result(string Status, long Total, long RestorationFee, string SubscriptionState);

        public interface ISubscriptionState
        {
            string CurrentStatus();
        }

        public interface INonprofitRegistry
        {
            bool IsEligible();
        }

        public interface IBillingRepository
        {
            void Save(Request request, Result result);
        }

        private readonly ISubscriptionState _subscriptions;
        private readonly INonprofitRegistry _nonprofits;
        private readonly IBillingRepository _billing;

        public SubscriptionService(ISubscriptionState subscriptions, INonprofitRegistry nonprofits, IBillingRepository billing)
        {
            _subscriptions = subscriptions ?? throw new ArgumentNullException(nameof(subscriptions), "Залежність не задано");
            _nonprofits = nonprofits ?? throw new ArgumentNullException(nameof(nonprofits), "Залежність не задано");
            _billing = billing ?? throw new ArgumentNullException(nameof(billing), "Залежність не задано");
        }

        public Result Process(Request request)
        {
            if (request is null)
                throw new ArgumentException("Заявку не задано", nameof(request));
            CheckRange(request.Users, 1, 49, "Кількість користувачів");
            CheckRange(request.Months, 1, 12, "Строк продовження");
            string state = _subscriptions.CurrentStatus();
            if (state != "Активна" && state != "Прострочена" && state != "Заблокована")
                throw new InvalidOperationException("Невідомий стан підписки");
            if (state == "Заблокована")
                return new Result("Підписку заблоковано", 0, 0, state);
            bool nonprofit = request.NonprofitClaimed && _nonprofits.IsEligible();
            int discount;
            if (nonprofit && request.AutoRenew)
                discount = 20;
            else if (nonprofit)
                discount = 20;
            else if (request.AutoRenew)
                discount = 5;
            else
                discount = 0;
            long basePrice = request.Users * request.Months * 20_000L;
            long total = basePrice - basePrice * discount / 100;
            if (request.Months >= 11)
                total = Math.Max(0, total - 10_000);
            long restorationFee = state == "Прострочена" ? 5_000 : 0;
            var result = new Result("Продовжено", total + restorationFee, restorationFee, "Активна");
            _billing.Save(request, result);
            return result;
        }

        private static void CheckRange(int value, int minimum, int maximum, string field)
        {
            if (value < minimum || value > maximum)
                throw new ArgumentException("Недопустиме значення: " + field);
        }
    }
}

File: group 2/variant11/subscription_service.py
from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True)
class Request:
    users: int
    months: int
    nonprofit_claimed: bool
    auto_renew: bool


@dataclass(frozen=True)
class Result:
    status: str
    total: int = 0
    restoration_fee: int = 0
    subscription_state: str | None = None


class SubscriptionState(Protocol):
    def current_status(self) -> str:
        ...


class NonprofitRegistry(Protocol):
    def is_eligible(self) -> bool:
        ...


class BillingRepository(Protocol):
    def save(self, request: Request, result: Result) -> None:
        ...


def _range(value: int, minimum: int, maximum: int, field: str) -> None:
    if type(value) is not int or not minimum <= value <= maximum:
        raise ValueError("Недопустиме значення: " + field)


def _flag(value: bool, field: str) -> bool:
    if type(value) is not bool:
        raise RuntimeError("Некоректні відомості: " + field)
    return value


def _count(value: int, field: str) -> int:
    if type(value) is not int or value < 0:
        raise RuntimeError("Некоректні відомості: " + field)
    return value


class SubscriptionService:
    def __init__(self, subscriptions: SubscriptionState, nonprofits: NonprofitRegistry, billing: BillingRepository) -> None:
        if subscriptions is None or nonprofits is None or billing is None:
            raise ValueError("Усі залежності мають бути задані")
        self._subscriptions = subscriptions
        self._nonprofits = nonprofits
        self._billing = billing

    def process(self, request: Request) -> Result:
        if not isinstance(request, Request):
            raise ValueError("Заявку не задано або її формат некоректний")
        _range(request.users, 1, 49, "Кількість користувачів")
        _range(request.months, 1, 12, "Строк продовження")
        if type(request.nonprofit_claimed) is not bool or type(request.auto_renew) is not bool:
            raise ValueError("Ознаки пільг мають бути логічними значеннями")
        state = self._subscriptions.current_status()
        if state not in ("Активна", "Прострочена", "Заблокована"):
            raise RuntimeError("Невідомий стан підписки")
        if state == "Заблокована":
            return Result("Підписку заблоковано", 0, 0, state)
        nonprofit = request.nonprofit_claimed and _flag(self._nonprofits.is_eligible(), "Неприбутковість")
        if nonprofit and request.auto_renew:
            discount = 20
        elif nonprofit:
            discount = 20
        elif request.auto_renew:
            discount = 5
        else:
            discount = 0
        base = request.users * request.months * 20_000
        total = base - base * discount // 100
        if request.months >= 11:
            total = max(0, total - 10_000)
        restoration_fee = 5_000 if state == "Прострочена" else 0
        result = Result("Продовжено", total + restoration_fee, restoration_fee, "Активна")
        self._billing.save(request, result)
        return result
